roachtest: Index overload automation of TPC-E

andrewbaptist opened the pull request:
This patch adds the ability to run tests against n2 standard from within roachprod / roachtest.

Release note: None
Epic: None

Commit:
roachprod: Add support for GCE n2-standard

This patch adds the ability to run tests against n2 standard
from within roachprod / roachtest.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90005)
<!-- Reviewable:end -->

Commit:
roachtest: Index overload using tpc-e

This introduces a test which runs an index creation. During the index
creation the latency of foreground traffic spikes even though it is not
actually using the index. The cause of this is contention for resources
between the foreground workload and the index creation at multiple
levels (CPU, IO, Pebble sublevels).

Release note: None

Epic: none

irfansharif reviewed (commented):
I know that @dt's working on adding roachtest libraries to snapshot GCE images and use them for quicker turn around times. He's also actually doing it for TPC-E restores I think. This PR should wait until that gets done -- this is far too long running and $$$ to be useful otherwise.

andrewbaptist commented:
What if we merged it with the skip test flag - since the ability to run skipped tests is about to merge

irfansharif commented on pkg/cmd/roachtest/test_registry.go:
```diff
@@ -137,7 +139,7 @@ func (r *testRegistryImpl) prepareSpec(spec *registry.TestSpec) error {
 	// tests should expect to take "less time". Longer-running tests require the
 	// weekly tag.
-	const maxTimeout = 18 * time.Hour
+	const maxTimeout = 7 * 24 * time.Hour
```
Why this change?

andrewbaptist replied:
Oops - removed this change. thanks for catching this.

irfansharif commented on pkg/cmd/roachtest/tests/admission_control_index_overload_tpce.go:
```diff
@@ -0,0 +1,172 @@
+	"github.com/cockroachdb/errors"
+)
+
+// TODO(baptist): This shares a lot of code with tpce.go. Consider ways to refactor.
```
Let's do this now.

irfansharif commented on pkg/cmd/roachtest/tests/admission_control_index_overload_tpce.go:
```diff
@@ -0,0 +1,172 @@
+			promCfg.WithPrometheusNode(workloadNode.InstallNodes()[0])
+			promCfg.WithNodeExporter(c.All().InstallNodes())
+			promCfg.WithCluster(roachNodes.InstallNodes())
+			promCfg.WithGrafanaDashboard("http://go.crdb.dev/p/snapshot-admission-control-grafana")
```
Use https://go.crdb.dev/p/index-admission-control-grafana instead.

andrewbaptist replied:
Done

irfansharif commented on pkg/cmd/roachtest/tests/admission_control_index_overload_tpce.go:
```diff
@@ -0,0 +1,172 @@
+
+			// Wait for the load to run for a while before starting the index build.
+			select {
+			case <-time.After(opts.delayBeforeIndex):
```
Look at how the TPC-C harness does things. It's very flexible, and better in many regards. There we have that `During` callback where test authors can inject whatever sleep they want during a test run, and this sort of thing would apply. So I would suggest improving the TPC-E test harness first as part of this PR and then adding a good use of it. It'll also encourage more TPC-E roachtests from other teams.

irfansharif commented on pkg/cmd/roachtest/tests/admission_control_index_overload_tpce.go:
```diff
@@ -0,0 +1,172 @@
+			t.Status("creating index (~30min)")
+
+			db := c.Conn(ctx, t.L(), 1)
+			_, err := db.ExecContext(ctx, "CREATE INDEX ON tpce.cash_transaction (ct_dts)")
```
The same suggestion as the other PR -- use an explicit, unique name, suffixed perhaps by the timestamp this was run at so that repeated runs are made easy.

andrewbaptist replied:
I realized when iterating with that one that not specifying a name gives the flexibility to re-run the identical create index (the system does this automatically) - so its not necessary to pick an explicit name unless you want to drop it in the end. Since I wasn't dropping I didn't pick a name. Let me know if you think I should change this.

irfansharif commented on pkg/cmd/roachtest/tests/admission_control_index_overload_tpce.go:
```diff
@@ -0,0 +1,172 @@
+	})
+
+	// This test takes about 1 hours to run.
+	r.Add(registry.TestSpec{
```
Meh, let's just have the one test above and make sure it runs <1h. And also, within reason, support either --local and/or --skip-init if you use an already set up cluster to keep re-running things without the bulk import. This TPC-E harness also needs to integrate with prometheus/grafana (I see you've done so manually above), let's do that in this PR -- steal as many ideas from the TPC-C harness when doing so.

andrewbaptist replied:
I'm not sure I understand this request. I don't think it is possible to create a test that runs <1hr, uses default settings and sees this overload issue (since it needs 1000 L0 files as part of index creation). I wasn't sure how best to approach this, so definitely open to ideas on how to do this (combination of skipped tests, short test that doesn't show the issue or other).

andrewbaptist replied:
Removed the small test below.

irfansharif reviewed (commented):
I'm ok merging this test as skipped until the GCE snapshotting work but we ought to improve the TPC-E harness first here instead of duplicating a bunch of code (even if this was the easiest thing to do at the start). Left some comments below but the TL;DR is lets make this as good as the TPC-C harness; I think it's doable since it's not a ton of code for either today. Feel free to do this work in whatever PRs/commits make life easiest for you.

irfansharif commented:
This is subsumed by https://github.com/cockroachdb/cockroach/pull/103816.
